### PR TITLE
Add constraints to tile dialect

### PR DESCRIPTION
**Pull request by flaub**

### CODE REVIEW

File: pmlc/dialect/tile/contraction.cc [hunk 1]
```diff
@@ -604,14 +613,17 @@ bool Contraction::NeedReduce() const {
...
   // Reduce if needed
   if (NeedReduce()) {
     ReduceOutputPolynomials(constraints);
     constraints = GatherConstraints(shapes);
   }
   constraints.MergeParallelConstraints();
+  IVLOG(3, "Merged Parallel Constraints:" << to_string(constraints.constraints));
   // Defract if needed (defract does early return if not required)
   Defractionalize(constraints);
+  // IVLOG(3, "Defracted:\n" << to_string(defracted));
```
Comment (earhart) [inline pmlc/dialect/tile/contraction.cc]: Optional: remove?  (Or uncomment?)

File: pmlc/dialect/tile/contraction.h [hunk 2]
```diff
@@ -48,7 +48,7 @@ struct Constraints {
 };
 
 struct Contraction {
-  explicit Contraction(ContractionOp op);
+  explicit Contraction(ContractionOp op, llvm::ArrayRef<ConstraintOp> constraintOps);
```
Comment (earhart) [inline pmlc/dialect/tile/contraction.h]: We usually only use `explicit` with single-parameter constructors -- are you trying to disallow copy-list initialization, or is this left over from previously?

File: pmlc/dialect/tile/lowering.cc [hunk 3]
```diff
@@ -625,7 +646,7 @@ OwningModuleRef LowerIntoStripe(ModuleOp workspace) {
     IVLOG(1, "LowerIntoStripe failed: " << mlir::debugString(*module->getOperation()));
     throw std::runtime_error("Lowering to stripe dialect failure");
   }
-  // IVLOG(1, "after:\n" << mlir::debugString(*module->getOperation()));
+  IVLOG(1, "after:\n" << mlir::debugString(*module->getOperation()));
```
Comment (earhart) [inline pmlc/dialect/tile/lowering.cc]: This seems like a slightly verbose logging level.

File: pmlc/dialect/tile/contraction.h (hunk 2 above)
Comment (flaub) [reply pmlc/dialect/tile/contraction.h]: Oops just leftover, thanks for catching this.

File: pmlc/dialect/tile/lowering.cc (hunk 3 above)
Comment (flaub) [reply pmlc/dialect/tile/lowering.cc]: Oops, meant to re-comment or remove it

**Pull request merged**